### PR TITLE
Splash Screen minimal show time. Issue #16264

### PR DESCRIPTION
**Pull request by mnesarco**
[Proposal] Small change to fix Splash Screen show time until a refactoring of StartupProcess is done.
Please consider for 1.0.

https://github.com/user-attachments/assets/0d759695-aa1a-4027-8fa0-698c6c5b12ed

### CODE REVIEW

File: src/Gui/MainWindow.cpp [hunk 1]
```diff
@@ -1917,13 +1917,15 @@ void MainWindow::startSplasher()
...
-            d->splashscreen = new SplashScreen(this->splashImage());
+            d->splashscreen = new SplashScreen(this->splashImage(), 
+                Qt::WindowStaysOnTopHint | Qt::SplashScreen | Qt::X11BypassWindowManagerHint);
 
             if (!hGrp->GetBool("ShowSplasherMessages", false)) {
                 d->splashscreen->setShowMessages(false);
             }
 
             d->splashscreen->show();
+            QTimer::singleShot(3000, this, SLOT(stopSplasher()));
```
Comment (chennes) [inline src/Gui/MainWindow.cpp:1928]: Can you update to use the current functor-based connection instead? Thanks!

Comment (mnesarco) [reply src/Gui/MainWindow.cpp]: Ok. changed.

**Comment by chennes** [comment]
Thanks - @FreeCAD/design-working-group is a three-second timer on the splashscreen OK for you?

**Comment by 3x380V** [comment]
This is basically the same as calling splash stop from `delayedStartup()` which made all toolbars gone on Wayland. Anyway, building to verify.

File: src/Gui/MainWindow.h [hunk 2]
```diff
@@ -140,8 +140,6 @@ class GuiExport MainWindow : public QMainWindow
     static MainWindow* getInstance();
     /** Starts the splasher at startup. */
     void startSplasher();
-    /** Stops the splasher after startup. */
-    void stopSplasher();
```
Comment (hyarion) [inline src/Gui/MainWindow.h:144]: You can move the method `void stopSplasher()` back to its original place since you changed the timer to a modern functor-based connection. No need to place it under Q_SLOTS.

Comment (mnesarco) [reply src/Gui/MainWindow.h]: Ok, done.

File: src/Gui/MainWindow.cpp [hunk 3]
```diff
@@ -1917,13 +1917,15 @@ void MainWindow::startSplasher()
             GetGroup("BaseApp")->GetGroup("Preferences")->GetGroup("General");
         // first search for an external image file
         if (hGrp->GetBool("ShowSplasher", true)) {
-            d->splashscreen = new SplashScreen(this->splashImage());
+            d->splashscreen = new SplashScreen(this->splashImage(), 
+                Qt::WindowStaysOnTopHint | Qt::SplashScreen | Qt::X11BypassWindowManagerHint);
```
Comment (3x380V) [inline src/Gui/MainWindow.cpp:1921]: `Qt::SplashScreen` is already set in [QSplashScreen](https://codebrowser.dev/qt5/qtbase/src/widgets/widgets/qsplashscreen.cpp.html#_ZN13QSplashScreenC1ERK7QPixmap6QFlagsIN2Qt10WindowTypeEE) constructor. Also there is a trailing whitespace added.

Comment (mnesarco) [reply src/Gui/MainWindow.cpp]: Ok, fixed

**Comment by 3x380V** [comment]
So I played with that a bit more. With current patch, I do not splash see screen at all. I need to extend time to 10s, then it shows reliably.
Btw, all this splash screen exercises are useful only to show how much startup process needs to revisit ;-)

**Comment by mnesarco** [comment]
@3x380V 10s is weird, too much, i wonder why, in my modest laptop 3s is enough to show reliably. You are right, the whole startup process needs to be reviewed, I was under illusion that a simple patch could make it tolerable for 1.0. without affecting other parts of the process.

**Comment by chennes** [comment]
@mnesarco you are right that for 1.0 we need a simple patch, and IMO this meets that requirement. We can discuss it in the review meeting on Monday, my inclination is to merge this into the 1.0 branch as-is, with the obvious caveat that we expect it to be very shortly superseded in 1.1dev.

**Comment by mnesarco** [comment]
I have an idea to improve 3secs reliability. I will try it tomorrow...

**Comment by 3x380V** [comment]
Attached is screencast with timer set to 10s. It is Qt6 build on Debian 12 using Wayland.

[Screencast from 2024-09-21 10-42-29.webm](https://github.com/user-attachments/assets/1d4ebfeb-50aa-4890-9dd4-9ea6944b9759)

My wild guess is splash screen is shown too late as wayland buffer is queued with other I/O and gets processed too late (my machine has classic HDD).

**Comment by 3x380V** [comment]
For the reference, here is how the [soonsplash](https://github.com/3x380V/FreeCAD/tree/soonsplash) attempt looks like. This is out of scope here as even preparatory #16438 is not suitable for 1.0, but perhaps you'll find something inside useful. The problem is not when we hide the splash, but when we are showing it. And that is too late in the process.

[Screencast from 2024-09-21 11-56-45.webm](https://github.com/user-attachments/assets/b916e524-48da-44f6-bd0e-e0805cb6aec9)

**Comment by mnesarco** [comment]
> Attached is screencast with timer set to 10s. It is Qt6 build on Debian 12 using Wayland.
> 
>  Screencast.from.2024-09-21.10-42-29.webm 
> My wild guess is splash screen is shown too late as wayland buffer is queued with other I/O and gets processed too late (my machine has classic HDD).

I see, but i don't have that environment to test. Isn't 1.0 targeted to Qt5 ?

**Comment by mnesarco** [comment]
> For the reference, here is how the [soonsplash](https://github.com/3x380V/FreeCAD/tree/soonsplash) attempt looks like. This is out of scope here as even preparatory #16438 is not suitable for 1.0, but perhaps you'll find something inside useful. The problem is not when we hide the splash, but when we are showing it. And that is too late in the process.
> 

I think that it is already established that there is no complete solution for 1.0. Any real solution will be post 1.0. My question is which platforms can benefit from the current temporal patch? I mean there are Windows, MacOS-intel, MacOS-Mx, Linux-X11, Linux-Wayland, ... In the current non patched version the splash screen shows only for a fraction of a second on all platforms if i am not wrong.

**Comment by 3x380V** [comment]
It is the same with Qt5 on wayland, perhaps even worse as FreeCAD triggers:
```
QOpenGLFunctions created with non-current context
(qt.qpa.wayland) Wayland does not support QWindow::requestActivate()
```
so it cannot raise window. This is irrelevant to your patch.
If you are interested in screencasting Qt5 build, I can do it, but there is no difference in this matter.

**Comment by mnesarco** [comment]
> so it cannot raise window. This is irrelevant to your patch. If you are interested in screencasting Qt5 build, I can do it, but there is no difference in this matter.

Please don't waste your time on it. I believe in your results. You are the wayland expert, i know nothing about wayland.

**Comment by mnesarco** [comment]
@chennes  I added a mechanism to ensure that the 3 seconds are counted after the EventLoop dispatches the previous events, it should avoid to set it to 10 seconds. It does not solve the other issues mentioned by @3x380V

**Comment by 3x380V** [comment]
@mnesarco Here's 9f65702aabd7ea2d4ada04c27ac876f8b309d88b screen cast, Qt5 build, Debian 12, Wayland.

[Screencast from 2024-09-21 16-45-56.webm](https://github.com/user-attachments/assets/5eed7c54-58dd-424f-bb91-9f4f23d04163)

**Comment by mnesarco** [comment]
> @mnesarco Here's [9f65702](https://github.com/FreeCAD/FreeCAD/commit/9f65702aabd7ea2d4ada04c27ac876f8b309d88b) screen cast, Qt5 build, Debian 12, Wayland.
> 
>  Screencast.from.2024-09-21.16-45-56.webm 
> Calling `QApplication::processEvents();` is no-op at the point you are calling it, just look at the [Qt source](https://codebrowser.dev/qt5/qtbase/src/corelib/kernel/qcoreapplication.cpp.html#_ZN16QCoreApplication13processEventsE6QFlagsIN10QEventLoop17ProcessEventsFlagEE).

@3x380V  Can you check if adding an eventloop has any impact on your side? it makes things apparently a lot faster on my side.

image

**Comment by 3x380V** [comment]
@mnesarco, it is strange it has any impact for you at all. If you look again at the source I pointed, you'll see current thread needs to have event dispatcher. You can look at it from another side: how could local `QEventLoop` interact with event you pushed from `show()` method? Everything already provided in 2fd8f36acf32f9cdd366307acc85ec0b7c4aac72

**Comment by mnesarco** [comment]
> @mnesarco, it is strange it has any impact for you at all. If you look again at the source I pointed, you'll see current thread needs to have event dispatcher. You can look at it from another side: how could local `QEventLoop` interact with event you pushed from `show()` method? Everything already provided in [2fd8f36](https://github.com/FreeCAD/FreeCAD/commit/2fd8f36acf32f9cdd366307acc85ec0b7c4aac72)

QEventLoop constructor calls `d->threadData.loadRelaxed()->ensureEventDispatcher();` so it ensures the current thread has an event dispatcher. At least it is how i understand it.

Anyway, I give up.

**Comment by mnesarco** [comment]
Just as a reference:


https://github.com/user-attachments/assets/c6533838-1e03-4980-9ad5-595819228d94



OS: Ubuntu 24.04.1 LTS
Word size of FreeCAD: 64-bit
Version: 1.1.0dev.38794 +1 (Git)
Build type: Unknown
Branch: splash-issue-16264
Hash: 89cb56e8e0066facb4d702a57ab2e9e5cee7a793
Python 3.12.3, Qt 5.15.13, Coin 4.0.2, Vtk 9.1.0, OCC 7.6.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: OpenDark.qss/OpenDark/Qt default
Installed mods: 
  * MnesarcoUtils 0.2.13
  * OpenTheme 2024.9.1

**Comment by 3x380V** [comment]
But [the constructor](https://codebrowser.dev/qt5/qtbase/src/corelib/kernel/qeventloop.cpp.html#_ZN10QEventLoopC1EP7QObject) needs `QCoreApplication` instance which does not exist at that point.

**Comment by mnesarco** [comment]
> But [the constructor](https://codebrowser.dev/qt5/qtbase/src/corelib/kernel/qeventloop.cpp.html#_ZN10QEventLoopC1EP7QObject) needs `QCoreApplication` instance which does not exist at that point.

Then why i don't see the warning: https://codebrowser.dev/qt5/qtbase/src/corelib/kernel/qeventloop.cpp.html#110

image

**Comment by 3x380V** [comment]
You are right here, `QCoreApplicationPrivate::threadRequiresCoreApplication()` is `false`, so [notifyInternal2](https://codebrowser.dev/qt5/qtbase/src/corelib/kernel/qcoreapplication.cpp.html#_ZN16QCoreApplication15notifyInternal2EP7QObjectP6QEvent) also works. I'll try to debug I/O queue once time permits, but there is probably not much we can do about it anyway.

**Comment by 3x380V** [comment]
You are right here. I was confused switching between your PR and my experimental branch. While I was able to hack splash to show almost at the beginning of `main()` there is still long path to clean it and I have no clue how to do it for 1.0.

**Comment by chennes** [comment]
Where does this stand? Obviously it is a stopgap solution for 1.0, but is it now ready to merge?

**Comment by mnesarco** [comment]
> Where does this stand? Obviously it is a stopgap solution for 1.0, but is it now ready to merge?

@chennes I think it is safe to merge as it is. It does not improve current behavior in wayland apparently, but others can benefit. It will break nothing anyway.

**Comment by 3x380V** [comment]
In fact it is much worse on Wayland for both Qt5 and Qt6 builds. With timer set to 3s, splash does not show at all. Probably depends largely on hardware. For the reference, current main **without** patch

[Screencast from 2024-09-22 21-17-01.webm](https://github.com/user-attachments/assets/a5d9750e-ec7d-41c8-a1d3-c0952cdb3e83)

**Comment by mnesarco** [comment]
@3x380V there is a possibility to make it not worse for wayland, i can update the PR so you can do a quick test. Do you want to try this little experiment?

**Comment by mnesarco** [comment]
I suppose that with the last change, 3secs can be reduced even to 2secs and it must not make waylan worse at least.

File: src/Gui/MainWindow.cpp [hunk 4]
```diff
@@ -1917,13 +1917,17 @@ void MainWindow::startSplasher()
...
+                Qt::WindowStaysOnTopHint | Qt::X11BypassWindowManagerHint);
 
             if (!hGrp->GetBool("ShowSplasherMessages", false)) {
                 d->splashscreen->setShowMessages(false);
             }
 
+            QEventLoop loop;
+            Q_UNUSED(loop)
             d->splashscreen->show();
+            QApplication::processEvents();
```
Comment (3x380V) [inline src/Gui/MainWindow.cpp:1930]: What about:
```
            QEventLoop loop;
            d->splashscreen->show();
            loop.processEvents(); 
```

**Comment by 3x380V** [comment]
> I suppose that with the last change, 3secs can be reduced even to 2secs and it must not make waylan worse at least.

Here's how it looks with 3s:

[Screencast from 2024-09-22 22-02-37.webm](https://github.com/user-attachments/assets/46864c02-ddc3-4df9-87a4-90c144294947)

**Comment by mnesarco** [comment]
> > I suppose that with the last change, 3secs can be reduced even to 2secs and it must not make waylan worse at least.
> 
> Here's how it looks with 3s:
> 
>  Screencast.from.2024-09-22.22-02-37.webm

Very weird, with the latest change it should behave like the non-patched version in the worst case.

File: src/Gui/MainWindow.cpp (hunk 4 above)
Comment (mnesarco) [reply src/Gui/MainWindow.cpp:1930]: Does it make any difference on your side?

**Comment by mnesarco** [comment]
> > I suppose that with the last change, 3secs can be reduced even to 2secs and it must not make waylan worse at least.
> 
> Here's how it looks with 3s:
> 
>  Screencast.from.2024-09-22.22-02-37.webm

Is it maybe possible that the splash screen in behind main window in your environment?

**Comment by 3x380V** [comment]
No, when I hide terminal, it looks the same. In theory, practice and theory are the same while in practice they are not ;-)
I may profile it later, but I have to finish some work till Tuesday, so I cannot spend much time on it now.
Also we may detect Wayland runtime and leave it as it is in that case.
Or we may just do it right and ask people to test extensively. Not sure about that, it is just a splash...

**Comment by kadet1090** [comment]
> Is it maybe possible that the splash screen is behind main window in your environment?

For wayland it would be the case: https://forum.qt.io/topic/143381/how-to-get-window-stays-on-top-in-plasma-wayland, as a workaround we can call `raise` on splashscreen but it may interfere with some stuff.

**Comment by 3x380V** [comment]
As noted previously for Qt5 and Wayland (which is target for 1.0) that does not work as it basically boils down to warning:
```
(qt.qpa.wayland) Wayland does not support QWindow::requestActivate()
```
Also splash is not hidden under other windows, it is (almost) not shown at all. Qt just renders shm buffer and sends it to render. That request is queued and executed when its turn comes. It is completely asynchronous to FreeCAD code and because startup process is I/O intensive, both requests are queued, but executed very soon one after another, thus splash could be shown for very short time under some circumstances. So you'd need to use a [wayland callback](https://wayland-book.com/surfaces-in-depth/frame-callbacks.html) and I'd honestly rather revisit whole startup process the choose this way.
In short, let's do it @mnesarco's way, but exclude Wayland as it makes things worse here.

**Comment by chennes** [comment]
@mnesarco can you add the proposed check for Wayland?

**Comment by mnesarco** [comment]
> @mnesarco can you add the proposed check for Wayland?

Sure, I will add the check in the `startSplasher()` because in `stopSplasher()` has no sense, but i can add the check there too if required.

**Comment by mnesarco** [comment]
@chennes @3x380V I have added a check for wayland in `startSplasher()` can you confirm if it is enough or do i need to also check in `stopSplasher()` please.

**Comment by mnesarco** [comment]
@chennes @3x380V updated. I changed the flags depending on wayland/others. Is is not exactly as it was before but may be enough, if not, then i can do the ugly change to ensure that it is exactly as before in case of wayland.

File: src/Gui/MainWindow.cpp [hunk 5]
```diff
@@ -1933,11 +1938,14 @@ void MainWindow::startSplasher()
...
-        d->splashscreen = nullptr;
-    }
+    QApplication::processEvents();
+    QTimer::singleShot(3000, this, [this]() {
+        if (d->splashscreen) {
+            d->splashscreen->finish(this);
+            d->splashscreen->deleteLater();
+            d->splashscreen = nullptr;
+        }
+    });
```
Comment (3x380V) [inline src/Gui/MainWindow.cpp:1948]: This change makes splash to be displayed for a very short time. However it would be nice if someone with fast permanent storage (SSD, NVMe) could test this as well.

Comment (mnesarco) [reply src/Gui/MainWindow.cpp:1948]: @3x380V  I can add the wayland case here to to make it works as non-patched version. It is still very difficult to understand why closing the splash immediately takes more time than closing it 3 seconds later. But as you said, theory and practice are not the same.

Comment (3x380V) [reply src/Gui/MainWindow.cpp:1948]: I guess the answer lays in `waitForWindowExposed()` implementation, but it is just a guess. With enough spare time it could probably be debugged using hardware timer, so we'd have more that a guesses.

Comment (mnesarco) [reply src/Gui/MainWindow.cpp]: @3x380V Ok, added the conditional there too.

**Comment by 3x380V** [comment]
@mnesarco, that's it, thank you!

**Comment by mnesarco** [comment]
Hey @chennes apparently we are ready to merge this. It will have a short life until someone has the time and knowledge to provide a definitive solution.

**Comment by chennes** [comment]
Thanks. Looks like that CI failure is fake, I will re-trigger it, it's just an HTTP glitch.
